lib/viewmodels/map/map_pin_view_model.dart

```diff
@@ -42,18 +57,66 @@ class MapPinViewModel extends AutoDisposeAsyncNotifier<List<MapPinDetails>> {
 
   /// Get nearby posts
   Future<List<MapPinDetails>> _getNearbyPosts() async {
-    final postRepository = ref.watch(postRepositoryServiceProvider);
     final position = await ref.watch(livePositionStreamProvider.future);
+    final postRepository = ref.watch(postRepositoryServiceProvider);
+    final userRepository = ref.watch(userRepositoryServiceProvider);
+    final userId = ref.watch(validLoggedInUserIdProvider);
 
     if (position == null) {
       return List.empty();
     }
 
+    final user = await userRepository.getUser(userId);
+
     final nearPosts = await postRepository.getNearPosts(
       position,
       PostsFeedViewModel.kmPostRadius,
     );
-    return nearPosts.map(_toMapPinDetails).toList();
+
+    return nearPosts
+        .map(
+          (post) => _toMapPinDetails(
+            post,
+            callback: context != null
+                ? () => showDialog(
+                      context: context!,
+                      builder: (context) {
+                        return MapPinPopUp(
+                          key: mapPinPopUpKey,
+                          title: post.data.title,
+                          content: post.data.description,
+                          displayButton: true,
+                          navigationAction: () {
+                            Navigator.pushNamed(
+                              context,
+                              Routes.post.name,
+                              arguments: PostDetails(
+                                postId: post.id,
+                                title: post.data.title,
+                                description: post.data.description,
+                                ownerCentauriPoints: user.data.centauriPoints,
+                                ownerDisplayName: user.data.displayName,
+                                ownerUsername: user.data.username,
+                                publicationDate:
+                                    post.data.publicationTime.toDate(),
+                                commentNumber: post.data.commentCount,
+                                voteScore: post.data.voteScore,
+                                distance:
+                                    (GeoFirePoint(position).distanceBetweenInKm(
+                                              geopoint: post.location.geoPoint,
+                                            ) *
+                                            1000)
+                                        .round(),
```

Comment:
There is quite a bit of code repetition between this and `posts_feed_view_model.dart` (lines 85 to 102). I think it would be interesting to add a factory `PostDetails.fromFirestoreData` which would take in parameter the `PostFiresore` post, the `UserFirestore` owner and the `GeoFirePoint` user position.

Reply:
Fixed with [5223934](https://github.com/ProximaEPFL/proxima/pull/335/commits/5223934e9ef70f7e76d529b65a3fd553cfbe35ca)